File: spine-libgdx/src/com/esotericsoftware/spine/attachments/MeshAttachment.java
/******************************************************************************
 * Spine Runtimes Software License
 * Version 2
 * 
 * Copyright (c) 2013, Esoteric Software
 * All rights reserved.
 * 
 * You are granted a perpetual, non-exclusive, non-sublicensable and
 * non-transferable license to install, execute and perform the Spine Runtimes
 * Software (the "Software") solely for internal use. Without the written
 * permission of Esoteric Software, you may not (a) modify, translate, adapt or
 * otherwise create derivative works, improvements of the Software or develop
 * new applications using the Software or (b) remove, delete, alter or obscure
 * any trademarks or any copyright, trademark, patent or other intellectual
 * property or proprietary rights notices on or in the Software, including
 * any copy thereof. Redistributions in binary or source form must include
 * this license and terms. THIS SOFTWARE IS PROVIDED BY ESOTERIC SOFTWARE
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
 * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL ESOTERIC SOFTARE BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *****************************************************************************/

package com.esotericsoftware.spine.attachments;

import com.esotericsoftware.spine.Bone;
import com.esotericsoftware.spine.Skeleton;
import com.esotericsoftware.spine.Slot;

import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.g2d.TextureRegion;
import com.badlogic.gdx.utils.FloatArray;
import com.badlogic.gdx.utils.NumberUtils;

/** Attachment that displays a texture region. */
public class MeshAttachment extends Attachment {
	private TextureRegion region;
	private String path;
	private float[] vertices;
	private short[] triangles;
	private float[] worldVertices;
	private final Color color = new Color(1, 1, 1, 1);

	// Nonessential.
	private int[] edges;
	private float width, height;
	private int hullLength;

	public MeshAttachment (String name) {
		super(name);
	}

	public void setRegion (TextureRegion region) {
		if (region == null) throw new IllegalArgumentException("region cannot be null.");
		this.region = region;
	}

	public TextureRegion getRegion () {
		if (region == null) throw new IllegalStateException("Region has not been set: " + this);
		return region;
	}

	public void updateWorldVertices (Slot slot, boolean premultipliedAlpha) {
		Skeleton skeleton = slot.getSkeleton();
		Color skeletonColor = skeleton.getColor();
		Color slotColor = slot.getColor();
		Color meshColor = color;
		float a = skeletonColor.a * slotColor.a * meshColor.a * 255;
		float multiplier = premultipliedAlpha ? a : 255;
		float color = NumberUtils.intToFloatColor( //
			((int)a << 24) //
				| ((int)(skeletonColor.b * slotColor.b * meshColor.b * multiplier) << 16) //
				| ((int)(skeletonColor.g * slotColor.g * meshColor.g * multiplier) << 8) //
				| (int)(skeletonColor.r * slotColor.r * meshColor.r * multiplier));

		float[] worldVertices = this.worldVertices;
		FloatArray verticesArray = slot.getAttachmentVertices();
		float[] vertices = this.vertices;
		if (verticesArray.size == vertices.length) vertices = verticesArray.items;
		Bone bone = slot.getBone();
		float x = skeleton.getX() + bone.getWorldX(), y = skeleton.getY() + bone.getWorldY();
		float m00 = bone.getM00(), m01 = bone.getM01(), m10 = bone.getM10(), m11 = bone.getM11();
		for (int v = 0, w = 0, n = worldVertices.length; w < n; v += 2, w += 5) {
			float vx = vertices[v];
			float vy = vertices[v + 1];
			worldVertices[w] = vx * m00 + vy * m01 + x;
			worldVertices[w + 1] = vx * m10 + vy * m11 + y;
			worldVertices[w + 2] = color;
		}
	}

	public float[] getWorldVertices () {
		return worldVertices;
	}

	public float[] getVertices () {
		return vertices;
	}

	public short[] getTriangles () {
		return triangles;
	}

	public Color getColor () {
		return color;
	}

	public String getPath () {
		return path;
	}

	public void setPath (String path) {
		this.path = path;
	}

	public int getHullLength () {
		return hullLength;
	}

	public void setHullLength (int hullLength) {
		this.hullLength = hullLength;
	}

	public int[] getEdges () {
		return edges;
	}

	public void setEdges (int[] edges) {
		this.edges = edges;
	}

	public float getWidth () {
		return width;
	}

	public void setWidth (float width) {
		this.width = width;
	}

	public float getHeight () {
		return height;
	}

	public void setHeight (float height) {
		this.height = height;
	}

	public void setMesh (float[] vertices, short[] triangles, float[] uvs) {
		this.vertices = vertices;
		this.triangles = triangles;

		int worldVerticesLength = vertices.length / 2 * 5;
		if (worldVertices == null || worldVertices.length != worldVerticesLength) worldVertices = new float[worldVerticesLength];

		float u, v, w, h;
		if (region == null) {
			u = v = 0;
			w = h = 1;
		} else {
			u = region.getU();
			v = region.getV();
			w = region.getU2() - u;
			h = region.getV2() - v;
		}
		for (int i = 0, ii = 3, n = vertices.length; i < n; i += 2, ii += 5) {
			worldVertices[ii] = u + uvs[i] * w;
			worldVertices[ii + 1] = v + uvs[i + 1] * h;
		}
	}
}

File: spine-cpp/src/spine/MeshAttachment.cpp
/******************************************************************************
 * Spine Runtimes License Agreement
 * Last updated April 5, 2025. Replaces all prior versions.
 *
 * Copyright (c) 2013-2025, Esoteric Software LLC
 *
 * Integration of the Spine Runtimes into software or otherwise creating
 * derivative works of the Spine Runtimes is permitted under the terms and
 * conditions of Section 2 of the Spine Editor License Agreement:
 * http://esotericsoftware.com/spine-editor-license
 *
 * Otherwise, it is permitted to integrate the Spine Runtimes into software
 * or otherwise create derivative works of the Spine Runtimes (collectively,
 * "Products"), provided that each user of the Products must obtain their own
 * Spine Editor license and redistribution of the Products in any form must
 * include this license and copyright notice.
 *
 * THE SPINE RUNTIMES ARE PROVIDED BY ESOTERIC SOFTWARE LLC "AS IS" AND ANY
 * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL ESOTERIC SOFTWARE LLC BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES,
 * BUSINESS INTERRUPTION, OR LOSS OF USE, DATA, OR PROFITS) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THE SPINE RUNTIMES, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *****************************************************************************/

#include <spine/MeshAttachment.h>
#include <spine/Atlas.h>

#include <assert.h>

using namespace spine;

RTTI_IMPL(MeshAttachment, VertexAttachment)

MeshAttachment::MeshAttachment(const String &name, Sequence *sequence)
	: VertexAttachment(name), _sequence(sequence), _regionUVs(), _triangles(), _hullLength(0), _path(), _color(1, 1, 1, 1), _sourceMesh(NULL),
	  _edges(), _width(0), _height(0) {
	assert(sequence);
}

MeshAttachment::~MeshAttachment() {
	delete _sequence;
}

void MeshAttachment::computeWorldVertices(Skeleton &skeleton, Slot &slot, size_t start, size_t count, float *worldVertices, size_t offset,
										  size_t stride) {
	VertexAttachment::computeWorldVertices(skeleton, slot, start, count, worldVertices, offset, stride);
}

Array<float> &MeshAttachment::getRegionUVs() {
	return _regionUVs;
}

void MeshAttachment::setRegionUVs(Array<float> &inValue) {
	_regionUVs.clearAndAddAll(inValue);
}

Array<unsigned short> &MeshAttachment::getTriangles() {
	return _triangles;
}

void MeshAttachment::setTriangles(Array<unsigned short> &inValue) {
	_triangles.clearAndAddAll(inValue);
}

int MeshAttachment::getHullLength() {
	return _hullLength;
}

void MeshAttachment::setHullLength(int inValue) {
	_hullLength = inValue;
}

Sequence &MeshAttachment::getSequence() {
	return *_sequence;
}

void MeshAttachment::updateSequence() {
	_sequence->update(*this);
}

const String &MeshAttachment::getPath() {
	return _path;
}

void MeshAttachment::setPath(const String &inValue) {
	_path = inValue;
}

Color &MeshAttachment::getColor() {
	return _color;
}

MeshAttachment *MeshAttachment::getSourceMesh() {
	return _sourceMesh;
}

void MeshAttachment::setSourceMesh(MeshAttachment *inValue) {
	_sourceMesh = inValue;
	if (inValue != NULL) {
		_bones.clearAndAddAll(inValue->_bones);
		_vertices.clearAndAddAll(inValue->_vertices);
		_regionUVs.clearAndAddAll(inValue->_regionUVs);
		_triangles.clearAndAddAll(inValue->_triangles);
		_hullLength = inValue->_hullLength;
		_worldVerticesLength = inValue->_worldVerticesLength;
		_edges.clearAndAddAll(inValue->_edges);
		_width = inValue->_width;
		_height = inValue->_height;
	}
}

Array<unsigned short> &MeshAttachment::getEdges() {
	return _edges;
}

void MeshAttachment::setEdges(Array<unsigned short> &inValue) {
	_edges.clearAndAddAll(inValue);
}

float MeshAttachment::getWidth() {
	return _width;
}

void MeshAttachment::setWidth(float inValue) {
	_width = inValue;
}

float MeshAttachment::getHeight() {
	return _height;
}

void MeshAttachment::setHeight(float inValue) {
	_height = inValue;
}

Attachment &MeshAttachment::copy() {
	if (_sourceMesh) return newLinkedMesh();

	MeshAttachment *copy = new (__FILE__, __LINE__) MeshAttachment(getName(), new (__FILE__, __LINE__) Sequence(*_sequence));
	copy->_path = _path;
	copy->_color.set(_color);
	copyTo(*copy);
	copy->_regionUVs.clearAndAddAll(_regionUVs);
	copy->_triangles.clearAndAddAll(_triangles);
	copy->_hullLength = _hullLength;
	copy->_edges.clearAndAddAll(_edges);
	copy->_width = _width;
	copy->_height = _height;
	return *copy;
}

MeshAttachment &MeshAttachment::newLinkedMesh() {
	MeshAttachment *copy = new (__FILE__, __LINE__) MeshAttachment(getName(), new (__FILE__, __LINE__) Sequence(*_sequence));
	copy->setTimelineAttachment(getTimelineAttachment());
	copy->_path = _path;
	copy->_color.set(_color);
	copy->setSourceMesh(_sourceMesh != NULL ? _sourceMesh : this);
	copy->updateSequence();
	return *copy;
}

void MeshAttachment::computeUVs(TextureRegion *region, Array<float> &regionUVs, Array<float> &uvs) {
	int n = (int) uvs.size();
	float u, v, width, height;
	if (region != NULL && region->getRTTI().instanceOf(AtlasRegion::rtti)) {
		AtlasRegion *r = static_cast<AtlasRegion *>(region);
		u = r->_u;
		v = r->_v;
		float textureWidth = r->getPage()->width;
		float textureHeight = r->getPage()->height;
		switch (r->_degrees) {
			case 90: {
				u -= (r->_originalHeight - r->_offsetY - r->_packedWidth) / textureWidth;
				v -= (r->_originalWidth - r->_offsetX - r->_packedHeight) / textureHeight;
				width = r->_originalHeight / textureWidth;
				height = r->_originalWidth / textureHeight;
				for (int i = 0; i < n; i += 2) {
					uvs[i] = u + regionUVs[i + 1] * width;
					uvs[i + 1] = v + (1 - regionUVs[i]) * height;
				}
				return;
			}
			case 180: {
				u -= (r->_originalWidth - r->_offsetX - r->_packedWidth) / textureWidth;
				v -= r->_offsetY / textureHeight;
				width = r->_originalWidth / textureWidth;
				height = r->_originalHeight / textureHeight;
				for (int i = 0; i < n; i += 2) {
					uvs[i] = u + (1 - regionUVs[i]) * width;
					uvs[i + 1] = v + (1 - regionUVs[i + 1]) * height;
				}
				return;
			}
			case 270: {
				u -= r->_offsetY / textureWidth;
				v -= r->_offsetX / textureHeight;
				width = r->_originalHeight / textureWidth;
				height = r->_originalWidth / textureHeight;
				for (int i = 0; i < n; i += 2) {
					uvs[i] = u + (1 - regionUVs[i + 1]) * width;
					uvs[i + 1] = v + regionUVs[i] * height;
				}
				return;
			}
			default: {
				u -= r->_offsetX / textureWidth;
				v -= (r->_originalHeight - r->_offsetY - r->_packedHeight) / textureHeight;
				width = r->_originalWidth / textureWidth;
				height = r->_originalHeight / textureHeight;
			}
		}
	} else if (region == NULL) {
		u = v = 0;
		width = height = 1;
	} else {
		u = region->_u;
		v = region->_v;
		width = region->_u2 - u;
		height = region->_v2 - v;
	}
	for (int i = 0; i < n; i += 2) {
		uvs[i] = u + regionUVs[i] * width;
		uvs[i + 1] = v + regionUVs[i + 1] * height;
	}
}
